### ISSUE TITLE
SDAG doesn't handle signed zero correctly when folding select -> {min|max}num

### ISSUE THREAD

**Issue by dtcxzyw** [llvm/llvm-project#93414]
Reproducer: https://godbolt.org/z/sqYqK8vd9

```
; llc -mtriple=riscv64 -mattr=+f test.ll -o -
define float @test(float %1) {
entry:
  %2 = fcmp olt float %1, 0.000000
  %3 = select i1 %2, float %1, float 0.000000
  ret float %3
}
```
```
test:                                   # @test
        fmv.w.x fa5, zero
        fmin.s  fa0, fa0, fa5
        ret
```

When `%1` is -0.0, the former returns +0.0 while the latter returns -0.0.
`SelectionDAGBuilder::visitSelect` incorrectly refined the select (always returns +0.0 when the input is -0.0) into `ISD::FMINNUM` (returns either 0.0 or -0.0).
Related code:
https://github.com/llvm/llvm-project/blob/a4a436672a2c179274e07aeb68e9acd6f483a653/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L3718-L3754
 
cc @topperc @arsenm

**Comment by dtcxzyw** [comment]
Original test case: https://github.com/dtcxzyw/llvm-codegen-benchmark/blob/main/dataset/75a2ca2dbd7ae03a.ll

**Comment by AZero13** [comment]
Is this issue still prevalent?

**Comment by arsenm** [comment]
ISD::FMINNUM was redefined to preserve signed 0 handling but not sure if the actual implementation was fixed

**Comment by jlebar** [comment]
Mythos 5 thinks this still reproduces.

Two more instances of this on NVPTX, both through the same SelectionDAGBuilder
fold, both in packed form:

```llvm
define <2 x float> @clamp_neg(<2 x float> %v) {
  %c = fcmp olt <2 x float> %v, zeroinitializer
  %r = select <2 x i1> %c, <2 x float> %v, <2 x float> zeroinitializer
  ret <2 x float> %r
}
```

`llc -march=nvptx64 -mcpu=sm_100 -mattr=+ptx86` emits `min.f32 %r3, %r2,
0f00000000` (x2, no fast-math needed — the constant arm makes the pattern
NaN-safe, so the fold fires on plain IR). For a `-0.0` lane the select must
produce `+0.0` (`olt(-0.0, +0.0)` is false, so the zeroinitializer arm is
taken), but `min` returns `-0.0` on hardware.

```llvm
define <2 x bfloat> @sel_max(<2 x bfloat> %a, <2 x bfloat> %b) {
  %c = fcmp nnan olt <2 x bfloat> %a, %b
  %r = select nnan <2 x i1> %c, <2 x bfloat> %b, <2 x bfloat> %a
  ret <2 x bfloat> %r
}
```

`llc -march=nvptx64 -mcpu=sm_90 -mattr=+ptx78` emits `max.bf16x2 %r3, %r1,
%r2`. For a lane with `a = -0.0, b = +0.0` the select must produce `a`
(`-0.0`), but `max.bf16x2` returns `+0.0`. `nnan` licenses assumptions about
NaNs, not about zero signs.

Both probed on hardware: the sign bit of the result differs from what the
unfolded compare+select produces. +1 to removing the fold (#93575).

**Comment by nikic** [comment]
Yes, there is a bug in the SPF logic in ValueTracking when it comes to signed zero handling. I've been working towards fixing that, and am nearly done with shaving this yak (still need to migrate s390x to pseudo_fmin/pseudo_fmax).
